text/0000-reformed-fn-pointers.md

```diff
@@ -0,0 +1,90 @@
+- Feature Name: reformed_fn_pointers
+- Start Date: 2015-03-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Make Rust's current function pointer types unsized, and introduce function reference types and new function pointer types, in order to make function references/pointers work more like value references/pointers.
+
+This is based on [RFC 883](https://github.com/rust-lang/rfcs/pull/883) and related discussions, where the following design is already largely agreed upon, but the author of RFC 883 doesn't have time to revise it. Therefore, this RFC is created to push for the changes.
+
+# Motivation
+
+Currently in Rust, there are two kinds of types that represent functions:
+
+1. Function item types: the types of function items, `fn(arg_list) -> ret_type {foo}`s.
+2. Function pointer types: the types of pointers to functions, `fn(arg_list) -> ret_type`s.
+
+Though called "function pointers", the function pointer types are considered *value* types (not reference/pointer types) by the language, which has the following problems:
+
+1. Inconsistencies in the language, especially when it comes to FFI codes that deal with nullable foreign function pointers. 
+2. There is currently no easy way to express the lifetime constraint on a function pointer to a non-`'static` function, which in particular makes it harder to implement type-safe JIT-ed functions/hot-loaded plugins in the future.
+
+Thus, this RFC proposes the following solution.
+
+# Detailed design
+
+Make the current function pointer types unsized, and introduce function reference types of the form `&fn(arg_list) -> ret_type` and new (const) function pointer types of the form `&const fn(arg_list) -> ret_type`.
+
+In the following section, `fn{f}`s denote function item types, `fn`s, `&fn`s and `*const fn`s denote current function pointer types, function reference types and new function pointer types, respectively. Those types are considered "compatible" if their `arg_list -> ret_type` parts match.
+
+The following rules will apply:
+
+1. `fn{f}`s are still the function item types.
+2. `fn`s are no longer function pointer types, but unsized types representing the bodies of functions.
+3. `&fn`s are function reference types, DST pointers with "auxiliary data" of type `()`.
```

Comment:
@petrochenkov, oops, I looked at the signature of `size_of_val` and didn't realize it should have been `T: ?Sized`.

@eddyb, I think  "`&fn` behaving the same" is okay as long as the resulting design is implementation-friendly and unsurprising to "external observers". "`fn{f}` being statically zero sized and `fn` being dynamically zero sized" is not a surprising design choice.

That said, "`fn{f}` being statically zero sized and `fn` being completely unsized" is not a surprising choice either. (Technical aspects aside, whether to choose one or the other depends a lot on how one wish to interpret the semantics of `fn`s. Are they "incomplete function items" or "function bodies"?)